From 8e4e2e7cf10ac1da7a6ce22c91f99d7cb91fb0e7 Mon Sep 17 00:00:00 2001
From: Alexandru Costache <alexandru@balena.io>
Date: Wed, 14 Dec 2022 16:51:36 +0100
Subject: [PATCH] lp55xx: Add support for default startup brightness from dtb

JF ticket https://jel.ly.fish/support-thread-1-0-0-front-cnv-ceqsqz1
requests for all d1-d8 of the LP55231 chip to be set from OS
startup to brightness = 255 and current = 10;

U-boot sets these values already but the led is turned off
as soon as the kernel driver is loaded. Turning it back on
once the container application starts is too late and adding
a systemd service would also cause a significant delay. Also,
setting the desired values from an udev rule would trigrer a
short flashing red color, which is noticeable.

We thus add support for using a default brightness from the
device-tree, and adapt it and the led_current to the requested
values provided by the customer in the above ticket.

Upstream-status: Inappropriate [configuration]
Signed-off-by: Alexandru Costache <alexandru@balena.io>
---
 drivers/leds/leds-lp55xx-common.c         | 6 +++++-
 include/linux/platform_data/leds-lp55xx.h | 1 +
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/leds/leds-lp55xx-common.c b/drivers/leds/leds-lp55xx-common.c
index 9fdfc1b9a1a0..a8027dd21a97 100644
--- a/drivers/leds/leds-lp55xx-common.c
+++ b/drivers/leds/leds-lp55xx-common.c
@@ -523,6 +523,9 @@ int lp55xx_register_leds(struct lp55xx_led *led, struct lp55xx_chip *chip)
 		/* setting led current at each channel */
 		if (cfg->set_led_current)
 			cfg->set_led_current(each, led_current);
+
+                each->brightness = pdata->led_config[i].default_brightness;
+                cfg->brightness_fn(each);
 	}
 
 	return 0;
@@ -575,7 +578,8 @@ static int lp55xx_parse_common_child(struct device_node *np,
 			    &cfg[led_number].led_current);
 	of_property_read_u8(np, "max-cur",
 			    &cfg[led_number].max_current);
-
+	if (of_property_read_u8(child, "default-brightness", &cfg[led_number].default_brightness))
+			cfg[i].default_brightness = 0;
 	ret = of_property_read_u32(np, "reg", chan_nr);
 	if (ret)
 		return ret;
diff --git a/include/linux/platform_data/leds-lp55xx.h b/include/linux/platform_data/leds-lp55xx.h
index 3441064713a3..efd38cd47ebf 100644
--- a/include/linux/platform_data/leds-lp55xx.h
+++ b/include/linux/platform_data/leds-lp55xx.h
@@ -28,6 +28,7 @@ struct lp55xx_led_config {
 	u8 chan_nr;
 	u8 led_current; /* mA x10, 0 if led is not connected */
 	u8 max_current;
+	u8 default_brightness;
 	int num_colors;
 	unsigned int max_channel;
 	int color_id[LED_COLOR_ID_MAX];
-- 
2.37.2

